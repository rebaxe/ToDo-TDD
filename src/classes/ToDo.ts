import FinishedTaskList from './FinishedTaskList'
import PointCounter from './PointCounter'
import Task from './Task'
import UnfinishedTaskList from './UnfinishedTaskList'

class ToDo {
  constructor(
    private unfinishedTaskList: UnfinishedTaskList,
    private finishedTaskList: FinishedTaskList,
    private pointCounter: PointCounter
  ) {}

  public getUnFinishedTasks(): UnfinishedTaskList {
    return this.unfinishedTaskList
  }

  public getFinishedTasks(): FinishedTaskList {
    return this.finishedTaskList
  }

  // public getPointCounter(): PointCounter {
  //   return this.pointCounter
  // }

  public toggleStatus(task: Task) {
    if (task.getStatus()) {
      task.setStatus(false)
      this.moveToUnfinishedList(task)
    } else {
      task.setStatus(true)
      this.moveToFinishedList(task)
    }
  }

  private moveToFinishedList(task: Task) {
    this.unfinishedTaskList.delete(task)
    this.finishedTaskList.add(task)
  }

  private moveToUnfinishedList(task: Task) {
    this.finishedTaskList.delete(task)
    this.unfinishedTaskList.add(task)
  }
}

export default ToDo
