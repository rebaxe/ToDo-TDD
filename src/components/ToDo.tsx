/* eslint-disable prettier/prettier */
import React, { ChangeEvent, FormEvent, useState } from 'react'
import EasyPoint from '../classes/EasyPoint'
import HardPoint from '../classes/HardPoint'
import MediumPoint from '../classes/MediumPoint'
import TaskClass from '../classes/Task'
import ToDoClass from '../classes/ToDo'

type Props = {
  todo: ToDoClass
}

function ToDo({ todo } : Props): JSX.Element {
  // const unfinishedTaskList = todo.getUnfinishedTasks()
  // const finishedTaskList = todo.getFinishedTasks()

  const [taskDescription, setTaskDescription] = useState('')
  const [taskPoint, setTaskPoint] = useState('')
  // const [unfinishedTasks, setUnfinishedTasks] = useState<Array<TaskClass>>(unfinishedTaskList.tasks)

  // const createPoint = () => {
  //   if (taskPoint === '1') {
  //     return new EasyPoint()
  //   }
  //   if (taskPoint === '2') {
  //     return new MediumPoint()
  //   } 
  //     return new HardPoint()
  //   }

  // const addPoint = () => {
  //   if (taskPoint) {
  //     const point = createPoint()
  //     const task = new TaskClass(taskDescription, false, point)
  //     unfinishedTaskList.add(task)
  //     setUnfinishedTasks([...unfinishedTaskList.tasks])
  //   }
  // }

  const handleNewTask = (e: FormEvent) => {
    e.preventDefault()
    setTaskDescription(taskDescription)
    // addPoint()
    // Efter submit ska inputfältet tömmas
    // setNewTask('')
  }  

  const handleSelectPoint = (e: ChangeEvent<HTMLInputElement>) => {
    setTaskPoint(e.target.value)
  }

  
 return (
   <div>
     <h1>Fun To Do</h1>
     <form onSubmit={e => handleNewTask(e)}>
       <input
         id="add-input"
         type="text"
         name="todo"
         onChange={e => setTaskDescription(e.target.value)}
         value={taskDescription}
         placeholder="Add new task..."
       />
       <button id="add-btn" type="submit" disabled={!taskPoint}>
         Add
       </button>
       <label htmlFor="easy">
        Easy
        <input type="radio" name="point" id="easy" value="1" onChange={e => handleSelectPoint(e)}/>
      </label>
      <label htmlFor="medium">
        Medium
        <input type="radio" name="point" id="medium" value="2" onChange={e => handleSelectPoint(e)}/>
      </label>
      <label htmlFor="hard">
        Hard
       <input type="radio" name="point" id="hard" value="3" onChange={e => handleSelectPoint(e)}/>
      </label>
     </form>
   </div>
 )
}

export default ToDo
