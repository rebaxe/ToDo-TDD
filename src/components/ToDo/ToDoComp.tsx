/* eslint-disable prettier/prettier */
import React, { ChangeEvent, FormEvent, useState } from 'react'
import EasyPoint from '../../classes/EasyPoint'
import HardPoint from '../../classes/HardPoint'
import MediumPoint from '../../classes/MediumPoint'
import TaskClass from '../../classes/Task'
import ToDoClass from '../../classes/ToDo'
import FinishedTasks from '../TaskLists/FinishedTasks'
import PointCounter from '../PointCounter/PointCounter'
import UnfinishedTasks from '../TaskLists/UnfinishedTasks'
import styles from './TodoComp.module.css'

type Props = {
  todo: ToDoClass
}

function ToDoComp({ todo } : Props): JSX.Element {
  const unfinishedTaskList = todo.getUnfinishedTasks()
  const finishedTaskList = todo.getFinishedTasks()
  const pointCounterClass = todo.getPointCounter()

  const [taskDescription, setTaskDescription] = useState('')
  const [taskPoint, setTaskPoint] = useState('')
  const [unfinishedTasks, setUnfinishedTasks] = useState<Array<TaskClass>>(unfinishedTaskList.getTasks())
  const [finishedTasks, setFinishedTasks] = useState<Array<TaskClass>>(finishedTaskList.getTasks())


  const createPoint = () => {
    if (taskPoint === '1') {
      return new EasyPoint()
    }
    if (taskPoint === '2') {
      return new MediumPoint()
    } 
    return new HardPoint()
  }

  const resetInputText = () => {
    setTaskDescription('')
  }

  const addPoint = () => {
    if (taskPoint) {
      const point = createPoint()
      const task = new TaskClass(taskDescription, false, point)
      unfinishedTaskList.add(task)
      setUnfinishedTasks([...unfinishedTaskList.getTasks()])
      resetInputText()
    }
  }

  const handleNewTask = (e: FormEvent) => {
    e.preventDefault()
    setTaskDescription(taskDescription)
    addPoint()
  }  

  const handleSelectPoint = (e: ChangeEvent<HTMLInputElement>) => {
    setTaskPoint(e.target.value)
  }

  const handleToggleStatus = (task: TaskClass) => {
    todo.toggleStatus(task)
    if (task.getStatus() === false) {
      setUnfinishedTasks([...unfinishedTaskList.getTasks()])
      setFinishedTasks([...finishedTaskList.getTasks()])
    } else {
      setUnfinishedTasks([...unfinishedTaskList.getTasks()])
      setFinishedTasks([...finishedTaskList.getTasks()])
    }
  }

  const handleDeleteTask = (task: TaskClass) => {
    if (task.getStatus() === false) {
      unfinishedTaskList.delete(task)
      setUnfinishedTasks([...unfinishedTaskList.getTasks()])
    } else {
      finishedTaskList.delete(task)
      setFinishedTasks([...finishedTaskList.getTasks()])
    }
  }

  const hasFinishedTasks = (): boolean => {
    return finishedTasks.length > 0
  }
  
  return (
    <div className={styles.todoContainer}>
      <div className={styles.todoHeadingWrapper}>
      <h1>Fun To Do</h1>
        <PointCounter pointCounter={pointCounterClass}/>
      </div>
      <div className={styles.taskListWrapper}>
        <UnfinishedTasks unfinishedTaskList={unfinishedTasks} handleToggleStatus={handleToggleStatus} hasFinishedTasks={hasFinishedTasks()} handleDeleteTask={handleDeleteTask} />
        {(!finishedTasks || finishedTasks.length < 1) ? null : <FinishedTasks finishedTaskList={finishedTasks} handleToggleStatus={handleToggleStatus} handleDeleteTask={handleDeleteTask} /> }
      </div>
      <div className={styles.addTaskForm} >
        <form onSubmit={e => handleNewTask(e)}>
          <div className={styles.formInputs}>
            <input
              id="add-input"
              type="text"
              name="todo"
              onChange={e => setTaskDescription(e.target.value)}
              value={taskDescription}
              placeholder="Add new task..."
            />
            <button id="add-btn" type="submit" disabled={!taskPoint || !taskDescription}>
              Add
            </button>
          </div>
          <div className="form-labels">
            <label htmlFor="easy">
              Easy
              <input type="radio" name="point" id="easy" value="1" onChange={e => handleSelectPoint(e)}/>
            </label>
            <label htmlFor="medium">
              Medium
              <input type="radio" name="point" id="medium" value="2" onChange={e => handleSelectPoint(e)}/>
            </label>
            <label htmlFor="hard">
              Hard
              <input type="radio" name="point" id="hard" value="3" onChange={e => handleSelectPoint(e)}/>
            </label>
          </div>
        </form>
      </div>
    </div>
  )
}

export default ToDoComp
