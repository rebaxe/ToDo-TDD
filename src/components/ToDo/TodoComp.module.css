:root {
  --color-orange-primary-opaque: #e76f5177;
}

.todoContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  position: relative;
  height: 100vh;
}

.todoHeadingWrapper {
  width: 100%;
  background-color: var(--color-teal);
  height: 150px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}

.taskListWrapper {
  width: 80%;
  height: auto;
  margin: 0 0 110px;
  overflow-y: scroll;
}

.addTaskForm {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 3% 0 3%;
  width: 100%;
  background-color: var(--color-teal);
  position: fixed;
  bottom: 0;
}

.formInputs {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  gap: 10px;
  margin: 5px;
}

.formLabels {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  gap: 5px;
}

.radioLabel {
  padding: 6px 8px;
  border-radius: 50px;
  display: inline-flex;
  cursor: pointer;
  transition: background 0.2s ease-in-out;
  margin: 8px 0;
  -webkit-tap-highlight-color: transparent;
  color: var(--color-offwhite);
}

.radioLabel:hover,
.radioLabel:focus-within {
  background: var(--color-orange-primary-opaque);
}

input[type='text'] {
  display: inline-block;
  padding: 10px 5px;
  border: 1px solid lightgray;
}

input[type='radio'] {
  vertical-align: middle;
  width: 20px;
  height: 20px;
  border-radius: 10px;
  background: none;
  border: 0;
  box-shadow: inset 0 0 0 1px var(--color-orange-primary);
  box-shadow: inset 0 0 0 1.5px var(--color-orange-primary);
  appearance: none;
  padding: 0;
  margin: 0 2px 0 0;
  transition: box-shadow 150ms cubic-bezier(0.95, 0.15, 0.5, 1.25);
  pointer-events: none;
}

input[type='radio']:focus {
  outline: none;
}

input[type='radio']:checked {
  box-shadow: inset 0 0 0 6px var(--color-orange-primary);
}

.labelSpan {
  vertical-align: middle;
  display: inline-block;
  line-height: 20px;
  padding: 0 8px;
}
