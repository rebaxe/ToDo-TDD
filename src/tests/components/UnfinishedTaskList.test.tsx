/* eslint-disable @typescript-eslint/no-explicit-any */
import React from 'react'
import { shallow } from 'enzyme'
import ToDo from '../../components/UnfinishedTasks'

let unfinishedTasks: any

// const simulateInputChange = (wrapper: any, inputSelector: any, newValue: any) => {
//   const input = wrapper.find(inputSelector)
//   input.simulate('change', {
//     target: { value: newValue }
//   })
//   return wrapper.find(inputSelector)
// }

describe('Unfinished Tasks component', () => {
  beforeEach(() => {
    unfinishedTasks = shallow(<UnfinishedTasks />)
  })

  it('Should render the unfinished task component', () => {
    expect(unfinishedTasks.getElements()).toMatchSnapshot()
  })
})
