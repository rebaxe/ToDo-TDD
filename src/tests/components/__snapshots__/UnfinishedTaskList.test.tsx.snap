// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unfinished Tasks component Should render the unfinished task component 1`] = `
Array [
  <React.Fragment>
    <p>
      No tasks
    </p>
  </React.Fragment>,
]
`;
